### PR TITLE
Add mtd interface for imxrt

### PR DESCRIPTION
**Pull request by HarrisonChoi750405**

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [e7eb5fed0f0c438d8d89edb1fbcf5220eedd8297] -

**Comment by seinfra** [comment]
Target : [e7eb5fed0f0c438d8d89edb1fbcf5220eedd8297] -

File: os/arch/arm/src/imxrt/imxrt_hyperflash.c [hunk 1]
```diff
@@ -478,12 +395,272 @@ static inline void flexspi_clock_init(void)
...
+	flashXfer.deviceAddress = addr * 2;
+	flashXfer.port = kFLEXSPI_PortA1;
+	flashXfer.cmdType = kFLEXSPI_Read;
+	flashXfer.SeqNumber = 1;
+	flashXfer.seqIndex = HYPERFLASH_CMD_LUT_SEQ_IDX_READDATA;
+	flashXfer.data = buffer;
+	flashXfer.dataSize = bytes;
+	status = imxrt_flexspi_transferblocking(base, &flashXfer);
+
+	if (status != kStatus_Success) {
```
Comment (kishore-sn) [inline os/arch/arm/src/imxrt/imxrt_hyperflash.c]: If the status is not equal to success, you have to print a debug message and return negative value.

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+	/* Send the "Bulk Erase (BE)" instruction */
+
+	(void)SPI_SEND(priv->dev, IMXRT_BE);
+
+	/* Deselect the FLASH */
+
+	SPI_SELECT(priv->dev, SPIDEV_FLASH, false);
+	#endif
+
+	fdbg("Return: OK\n");
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fvdbg

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+/************************************************************************************
+ * Name: imxrt_erase
+ ************************************************************************************/
+
+static int imxrt_erase(FAR struct mtd_dev_s *dev, off_t startsector, size_t nsectors)
+{
+	FAR struct imxrt_dev_s *priv = (FAR struct imxrt_dev_s *)dev;
+	size_t sectors_erased = 0;
+
+	fdbg("startsector: %d nsectors: %d\n", (long)startsector, (int)nsectors);
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fvdbg

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+}
+
+/************************************************************************************
+ * Name: imxrt_read
+ ************************************************************************************/
+
+static ssize_t imxrt_read(FAR struct mtd_dev_s *dev, off_t offset, size_t nbytes, FAR uint8_t *buffer)
+{
+	status_t status;
+	//	FAR struct imxrt_dev_s *priv = (FAR struct imxrt_dev_s *)dev;
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: How about deleting this comment?

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+	fvdbg("offset: %08lx nbytes: %d\n", (long)offset, (int)nbytes);
+
+	/* We must test if the offset + count crosses one or more pages
+	 * and perform individual writes.  The devices can only write in
+	 * page increments.
+	 */
+
+	startpage = offset / (1 << priv->pageshift);
+	endpage = (offset + nbytes) / (1 << priv->pageshift);
+//	startpage = offset / (1 << priv->sectorshift);
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: How about deleting this comment (+below line also)?

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+	if (startpage == endpage) {
+		/* All bytes within one programmable page.  Just do the write. */
+
+		imxrt_bytewrite(priv, buffer, offset, nbytes);
+	} else {
+		/* Write the 1st partial-page */
+
+		count = nbytes;
+		pagesize = (1 << priv->pageshift);
+//		pagesize = (1 << priv->sectorshift);
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: How about deleting this line..?

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 7]
```diff
@@ -0,0 +1,673 @@
...
+ * Name: imxrt_ioctl
+ ************************************************************************************/
+
+static int imxrt_ioctl(FAR struct mtd_dev_s *dev, int cmd, unsigned long arg)
+{
+	FAR struct imxrt_dev_s *priv = (FAR struct imxrt_dev_s *)dev;
+	int ret = -EINVAL;			/* Assume good command with bad parameters */
+	char tempLog[128];
+
+	fdbg("cmd: %d \n", cmd);
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fvdbg

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 8]
```diff
@@ -0,0 +1,673 @@
...
+			}
+
+			ret = OK;
+
+			snprintf(tempLog, sizeof(tempLog), "sectorshift: %d nsectors: %d pageshift: %d, npages: %d\n", priv->sectorshift, priv->nsectors, priv->pageshift, priv->npages);
+			IMXLOG(tempLog);
+
+			snprintf(tempLog, sizeof(tempLog), "blocksize: %d erasesize: %d neraseblocks: %d\n", geo->blocksize, geo->erasesize, geo->neraseblocks);
+			IMXLOG(tempLog);
+			fdbg("blocksize: %d erasesize: %d neraseblocks: %d\n", geo->blocksize, geo->erasesize, geo->neraseblocks);
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fvdbg

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 9]
```diff
@@ -0,0 +1,673 @@
...
+	}
+	break;
+
+	case MTDIOC_XIPBASE:
+	default:
+		ret = -ENOTTY;			/* Bad command */
+		break;
+	}
+
+	fdbg("return %d\n", ret);
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fvdbg

File: os/arch/arm/src/imxrt/imxrt_norflash.c [hunk 10]
```diff
@@ -304,22 +305,29 @@ const uint32_t customLUT[CUSTOM_LUT_LENGTH] = {
...
  *
  ************************************************************************************/
-status_t flexspi_nor_write_enable(FLEXSPI_Type *base, uint32_t baseAddr)
+status_t imxrt_flexspi_nor_write_enable(FLEXSPI_Type *base, uint32_t baseAddr)
 {
 	flexspi_transfer_t flashXfer;
 	status_t status;
 
-	/* Write enable */
+	/* Write neable */
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_norflash.c]: enable?

**Comment by chanijjani** [comment]
@HarrisonChoi750405  And please update the .config I sent to you by e-mail into TizenRT/build/configs/imxrt1050-evk/nxp_demo/defconfig also into this PR.

File: os/arch/arm/src/imxrt/imxrt_norflash.c (hunk 10 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_norflash.c]: fixed

**Comment by seinfra** [comment]
Target : [72e3ecb55ec056c7932b47f5dfd2f25cecf85aa5] - Code Rule Check (C++) OK.

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c (hunk 9 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

**Comment by seinfra** [comment]
Target : [72e3ecb55ec056c7932b47f5dfd2f25cecf85aa5] - Code Rule Check OK.

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c (hunk 8 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c (hunk 7 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c (hunk 6 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 11]
```diff
@@ -0,0 +1,673 @@
...
+
+static void imxrt_lock(FAR struct spi_dev_s *dev)
+{
+}
+
+/************************************************************************************
+ * Name: imxrt_unlock
+ ************************************************************************************/
+
+static inline void imxrt_unlock(FAR struct spi_dev_s *dev)
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 12]
```diff
@@ -0,0 +1,673 @@
...
+
+
+
+
+
+/************************************************************************************
+ * Name: imxrt_lock
+ ************************************************************************************/
+
+static void imxrt_lock(FAR struct spi_dev_s *dev)
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 13]
```diff
@@ -0,0 +1,673 @@
...
+static int imxrt_ioctl(FAR struct mtd_dev_s *dev, int cmd, unsigned long arg);
+
+/************************************************************************************
+ * Private Data
+ ************************************************************************************/
+
+/************************************************************************************
+ * Private Functions
+ ************************************************************************************/
+
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_hyperflash.c [hunk 14]
```diff
@@ -478,12 +395,272 @@ static inline void flexspi_clock_init(void)
...
+		return status;
+	}
+
+	status = imxrt_flexspi_nor_wait_bus_busy(base);
+
+	arch_flush_dcache(address + IMXRT_FLASH_BASE, address + IMXRT_FLASH_BASE + IMXRT_FLASH_PAGE_SIZE);
+
+	//flexspi_clock_update();
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_hyperflash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_hyperflash.c [hunk 15]
```diff
@@ -478,12 +395,272 @@ static inline void flexspi_clock_init(void)
...
+	flashXfer.dataSize = IMXRT_FLASH_PAGE_SIZE;
+	status = imxrt_flexspi_transferblocking(base, &flashXfer);
+
+	if (status != kStatus_Success) {
+		return status;
+	}
+
+	status = imxrt_flexspi_nor_wait_bus_busy(base);
+
+	arch_flush_dcache(address + IMXRT_FLASH_BASE, address + IMXRT_FLASH_BASE + IMXRT_FLASH_PAGE_SIZE);
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_hyperflash.c]: This operation is necessary because there is in SDK sample code

**Comment by chanijjani** [comment]
@HarrisonChoi750405 I'll submit a PR that has the update of "TizenRT/build/configs/imxrt1050-evk/nxp_demo/defconfig" also.
Just resolve a conflict please.

**Comment by HarrisonChoi750405** [comment]
Will close due to conflict

**Pull request closed without merging**